backend/cfg/cfg_dominators.ml

```diff
@@ -72,49 +72,121 @@ let invariant_doms : Cfg.t -> doms -> unit =
           doms))
     doms
 
-(* CR-soon xclerc for xclerc: factor out with the function in
-   `Regalloc_ls_utils`; the only difference is the point where `f` is called. *)
-let iter_blocks_dfs : Cfg.t -> f:(Cfg.basic_block -> unit) -> unit =
- fun cfg ~f ->
-  let marked = ref Label.Set.empty in
-  let rec iter (label : Label.t) : unit =
-    if not (Label.Set.mem label !marked)
+module DFSUtils : sig
+  type component
+  (* Identifier of a "component", which is the set of blocks reachable from a
+     given (pseudo-)entry (i.e. a block with no predecessors). *)
+
+  val equal_component : component -> component -> bool
+
+  type components
+  (* Map from block labels to component identifier. *)
+
+  val find_component : components -> Label.t -> component
+  (* [find_component components label] return the component identifier for
+     [label] in [components] - cannot raise. *)
+
+  val iter_blocks : Cfg.t -> f:(Cfg.basic_block -> unit) -> components
+  (* [iter_blocks cfg ~f] iterates over the blocks of [cfg] in depth-first
+     order, calling [f] on each node. In order to visit all blocks, even the
+     ones actually dead, it first iterates from the entry point of [cfg], and
+     then does the same from each pseudo-entry. The entry, and then each
+     pseudo-entry, defines a component (set of blocks reachable from that
+     origin), and the returned value is a (total) map from block labels to
+     components. *)
+end = struct
+  type component = int
+
+  let equal_component = Int.equal
+
+  type components = component Label.Tbl.t
+
+  let find_component components label =
+    match Label.Tbl.find_opt components label with
+    | None -> Misc.fatal_errorf "no component identifier for label %d" label
+    | Some component -> component
+
+  let iter_blocks_from :
+      Cfg.t ->
+      from:Label.t ->
+      f:(Cfg.basic_block -> unit) ->
+      component:component ->
+      components:components ->
+      unit =
+   fun cfg ~from ~f ~component ~components ->
+    let rec iter (label : Label.t) : unit =
+      if not (Label.Tbl.mem components label)
+      then (
+        Label.Tbl.replace components label component;
+        let block = Cfg.get_block_exn cfg label in
+        Label.Set.iter
+          (fun succ_label -> iter succ_label)
+          (Cfg.successor_labels ~normal:true ~exn:true block);
+        f block)
+    in
+    iter from
+
+  exception Found of Label.t
+
+  let find_pseudo_entry : Cfg.t -> components:components -> Label.t =
+   fun cfg ~components ->
+    try
+      Cfg.iter_blocks cfg ~f:(fun label block ->
+          if (not (Label.Tbl.mem components label))
+             && Label.Set.is_empty block.predecessors
+          then raise (Found label));
+      Misc.fatal_error "did not find a block with no predecessors"
+    with Found label -> label
+
+  let rec iter_blocks_pseudo_entries :
```

Comment:
This `iter_blocks_pseudo_entries` is more complicated than I expected. We can find all blocks with no predecessors  using `Cfg.iter_blocks`, and then just trace from there, using `iter_blocks_from`.  (This won't find unreachable cycles but neigher the current code does). Am I missing something?

Reply:
I don't really follow; how does that differ
from what is implemented?

Reply:
(We are not finding unreachable cycles,
because the assumption is that they do
not exist. But if they do, the program will
fail with `Misc.fatal_error "did not find
a block with no predecessors"`.)

Reply:
> I don't really follow; how does that differ from what is implemented?

current version finds pseudo_entry blocks one at a time, and iterates from it.  what I was describing is to first find them all (put in a temporary set), and then iterate from them - I think it is simpler and no need for a recursive function.

> (We are not finding unreachable cycles,
because the assumption is that they do
not exist. But if they do, the program will
fail with Misc.fatal_error "did not find a block with no predecessors".)

aha, that's what I was missing! and perhaps that's why you used this more complicated recursion, but I would still prefer the above and a check at the end if not all labels have components, then fail. i think it would work too and would be simpler to understand because of the separate concerns.

Reply:
ah, thanks for the clarification, I thought
it was more obvious correct by kind of
peeling what remains component per
component but maybe not.

Reply:
@gretay-js Let's go with the current version for now, as this is really blocking.  If we want to restructure the recursion, then that could be considered for a future PR.